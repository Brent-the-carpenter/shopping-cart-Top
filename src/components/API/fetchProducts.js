export const fetchProducts = async () => {
  try {
    const response = await fetch("https://fakestoreapi.com/products", {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
      },
      mode: "cors",
    });
    if (!response.ok) {
      const errorBody = await response.text();
      throw new Error(
        `HTTP error ${response.status}: ${response.statusText} - ${errorBody}`
      );
    }

    const data = await response.json();
    return data;
  } catch (error) {
    throw error;
  }
};
