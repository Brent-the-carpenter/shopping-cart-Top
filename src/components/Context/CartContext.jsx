import { createContext, useContext, useState } from "react";

const CartContext = createContext(null);
export const useCartContext = () => useContext(CartContext);

export const CartProvider = ({ children }) => {
  const [cart, setCart] = useState([]);

  const addToCart = (item) => {
    setCart((currentCart) => {
      const foundIndex = currentCart.findIndex(
        (cartItem) => cartItem.id === item.id
      );
      if (foundIndex !== -1) {
        const newCart = [...currentCart];
        newCart[foundIndex] = {
          ...newCart[foundIndex],
          quantity: newCart[foundIndex].quantity + 1,
        };
        return newCart;
      }

      return [...currentCart, { ...item, quantity: item.quantity || 1 }];
    });
  };

  const removeFromCart = (item) => {
    setCart((currentCart) =>
      currentCart.filter((cartItem) => cartItem.id !== item.id)
    );
  };

  const increaseQuantity = (item) => {
    setCart((currentCart) =>
      currentCart.map((cartItem) =>
        cartItem.id === item.id
          ? { ...cartItem, quantity: cartItem.quantity + 1 }
          : cartItem
      )
    );
  };

  const decreaseQuantity = (item) => {
    setCart((currentCart) =>
      currentCart.map((cartItem) => {
        if (cartItem.id === item.id) {
          const newQuantity = cartItem.quantity - 1;
          return newQuantity > 0
            ? { ...cartItem, quantity: newQuantity }
            : cartItem;
        }
        return cartItem;
      })
    );
  };

  const clearCart = () => {
    setCart([]);
  };

  const contextValue = {
    cart,
    addToCart,
    removeFromCart,
    clearCart,
    increaseQuantity,
    decreaseQuantity,
  };

  return (
    <CartContext.Provider value={contextValue}>{children}</CartContext.Provider>
  );
};
