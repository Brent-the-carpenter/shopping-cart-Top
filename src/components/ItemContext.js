import { createContext, useContext } from "react";

const ItemContext = createContext(null);
export const useItemsContext = () => useContext(ItemContext);

export { ItemContext };
