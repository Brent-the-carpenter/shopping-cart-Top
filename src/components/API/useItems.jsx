import { useState, useEffect } from "react";
import { fetchProducts } from "./fetchProducts";
const useItems = () => {
  const [items, setItems] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(null);
  useEffect(() => {
    let isCancelled = false;
    const fetchData = async () => {
      try {
        const data = await fetchProducts();
        if (!isCancelled) {
          setItems(data);
          setError(null);
          console.log(data);
        }
      } catch (error) {
        if (!isCancelled) {
          error.message === ""
            ? setError("An error occurred")
            : setError(error);
          setItems([]);
          console.log(error.message);
        }
      } finally {
        if (!isCancelled) {
          setLoading(false);
        }
      }
    };
    fetchData();
    return () => {
      isCancelled = true;
    };
  }, []);
  return { items, loading, error };
};
export { useItems };
