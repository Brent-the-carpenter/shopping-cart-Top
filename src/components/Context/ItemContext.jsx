import React, { createContext, useContext, useMemo } from "react";
import { useItems } from "../API/useItems";

const ItemContext = createContext({
  items: [],
  loading: true,
  error: null,
});

export const useItemsContext = () => useContext(ItemContext);

export const ItemProvider = ({ children }) => {
  const { items, loading, error } = useItems();

  const value = useMemo(
    () => ({
      items,
      loading,
      error,
    }),
    [items, loading, error]
  );

  return <ItemContext.Provider value={value}>{children}</ItemContext.Provider>;
};
