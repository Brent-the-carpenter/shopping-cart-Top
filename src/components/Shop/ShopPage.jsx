import React, { useState, useEffect, useMemo, useCallback } from "react";
import { useItemsContext } from "../Context/ItemContext";
import Item from "./Item";
import ShopContainer from "./SyledPage";
import debounce from "lodash.debounce";

function ShopPage() {
  const { items, loading, error } = useItemsContext();
  const [search, setSearch] = useState("");
  const [filteredItems, setFilteredItems] = useState(items);

  const debouncedSearch = useMemo(
    () =>
      debounce((searchTerm) => {
        const filtered = items.filter((item) =>
          item.title.toLowerCase().includes(searchTerm.toLowerCase())
        );
        setFilteredItems(filtered);
      }, 300),
    [items]
  );

  const handleSearch = useCallback(
    (e) => {
      const searchTerm = e.target.value;
      setSearch(searchTerm);
      debouncedSearch(searchTerm);
    },
    [debouncedSearch]
  );

  useEffect(() => {
    setFilteredItems(items);
  }, [items]);

  if (loading) return <h1>Loading...</h1>;
  if (error) return <h1>Error: {error.message}</h1>;

  return (
    <ShopContainer>
      <h1>Shop Page</h1>
      <div className="search-container">
        <input
          onChange={handleSearch}
          type="text"
          placeholder="Search..."
          value={search}
        />
        <svg
          className="SearchIcon"
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
        >
          <path d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" />
        </svg>
      </div>
      {filteredItems.length === 0 ? (
        <h1 id="not-found">No items found</h1>
      ) : (
        <div className="item-container">
          {filteredItems.map((item) => (
            <Item className="item" key={item.id} item={item} />
          ))}
        </div>
      )}
    </ShopContainer>
  );
}

export default ShopPage;
