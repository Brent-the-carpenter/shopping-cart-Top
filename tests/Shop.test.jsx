import { render, screen, waitFor } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import userEvent from "@testing-library/user-event";
import { BrowserRouter } from "react-router-dom";
import { CartProvider } from "../src/components/Context/CartContext";
import { ItemProvider } from "../src/components/Context/ItemContext";
import ShopPage from "../src/components/Shop/ShopPage";
import * as useItemsModule from "../src/components/API/useItems";

// Mock the entire module
vi.mock("../src/components/API/useItems");

describe("Shop Component", () => {
  beforeEach(() => {
    // Reset mocks and define default behavior
    vi.resetAllMocks();
    useItemsModule.useItems.mockReturnValue({
      items: [
        {
          id: 1,
          title: "Item 1",
          price: 10,
          description: "Test Description",
          image: "/path/to/image.jpg",
        },
        {
          id: 2,
          title: "Item 2",
          price: 20,
          description: "Test Description 2",
          image: "/path/to/image2.jpg",
        },
      ],
      loading: false,
      error: null,
    });
  });

  it("displays items with all details when not loading and no error", async () => {
    const { getByRole, getAllByText, getByTestId } = render(
      <BrowserRouter>
        <CartProvider>
          <ItemProvider>
            <ShopPage />
          </ItemProvider>
        </CartProvider>
      </BrowserRouter>
    );

    await waitFor(() => {
      expect(getByRole("heading", { name: "Item 1" })).toBeInTheDocument();
    });

    const button = getByTestId("show-button-1");
    userEvent.click(button);

    await waitFor(() => {
      expect(getAllByText("Test Description")).toHaveLength(1);
      expect(getAllByText("Add to Cart")).toHaveLength(2);
    });

    expect(getByTestId("shop-container")).toMatchSnapshot();
  });
});

describe("Shop Component Error Handling", () => {
  beforeEach(() => {
    useItemsModule.useItems.mockReturnValue({
      items: [],
      loading: false,
      error: { message: "Not found" },
    });
  });

  it("displays error message when error occurs", async () => {
    render(
      <BrowserRouter>
        <CartProvider>
          <ItemProvider>
            <ShopPage />
          </ItemProvider>
        </CartProvider>
      </BrowserRouter>
    );

    await waitFor(() => {
      expect(screen.getByText("Error: Not found")).toBeInTheDocument();
    });
  });
});

describe("Shop Component Loading State", () => {
  beforeEach(() => {
    useItemsModule.useItems.mockReturnValue({
      items: [],
      loading: true,
      error: null,
    });
  });

  it("displays loading state when loading", async () => {
    render(
      <BrowserRouter>
        <CartProvider>
          <ItemProvider>
            <ShopPage />
          </ItemProvider>
        </CartProvider>
      </BrowserRouter>
    );

    await waitFor(() => {
      expect(screen.getByText("Loading...")).toBeInTheDocument();
    });
  });
});
