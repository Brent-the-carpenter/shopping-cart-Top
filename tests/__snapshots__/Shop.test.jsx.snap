// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop Component > displays items with all details when not loading and no error 1`] = `
<div
  class="ShopContainer-sc-zy8dpt-0 exddvK"
  data-testid="shop-container"
>
  <h1>
    Shop Page
  </h1>
  <div
    class="search-container"
  >
    <input
      placeholder="Search..."
      type="text"
      value=""
    />
    <svg
      class="SearchIcon"
      data-testid="search-icon"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z"
      />
    </svg>
  </div>
  <div
    class="item-container"
  >
    <div
      class="Div-sc-1chur02-0 kzhfk item"
    >
      <div
        class="image-container"
      >
        <img
          alt="Item 1"
          src="/path/to/image.jpg"
        />
      </div>
      <h2>
        Item 1
      </h2>
      <div
        class="description"
      >
         
        Description
        <button
          data-testid="show-button-1"
          id="show-button"
        >
          <svg
            class=""
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M7,10L12,15L17,10H7Z"
            />
          </svg>
        </button>
      </div>
      <p
        class=""
      >
        Test Description
      </p>
      <div
        class="action-container"
      >
        <div>
          $
          10
        </div>
        <button
          data-testid="add-to-cart-1"
        >
          Add to Cart
        </button>
      </div>
    </div>
    <div
      class="Div-sc-1chur02-0 kzhfk item"
    >
      <div
        class="image-container"
      >
        <img
          alt="Item 2"
          src="/path/to/image2.jpg"
        />
      </div>
      <h2>
        Item 2
      </h2>
      <div
        class="description"
      >
         
        Description
        <button
          data-testid="show-button-2"
          id="show-button"
        >
          <svg
            class=""
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0,0 22,12A10,10 0 0,0 12,2M12,4A8,8 0 0,1 20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4M7,10L12,15L17,10H7Z"
            />
          </svg>
        </button>
      </div>
      <p
        class=""
      >
        Test Description 2
      </p>
      <div
        class="action-container"
      >
        <div>
          $
          20
        </div>
        <button
          data-testid="add-to-cart-2"
        >
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
